src/multisig/multisig_kex_msg.cpp

```diff
@@ -0,0 +1,236 @@
+
+    // decode base58 part
+    std::string decoded;
+    CHECK_AND_ASSERT_THROW_MES(tools::base58::decode(m_msg.substr(MULTISIG_KEX_MSG_V2_MAGIC.size()), decoded),
+      "Multisig kex msg decoding error.");
+
+    /// parse message
+    CHECK_AND_ASSERT_THROW_MES(decoded.size() >=
+      sizeof(std::uint32_t) + sizeof(crypto::public_key) + sizeof(crypto::signature), "Multisig kex msg malformed.");
+
+    // 1. signed message
+    std::string signed_msg = MULTISIG_KEX_MSG_V2_MAGIC;
+    signed_msg += decoded.substr(0, decoded.size() - sizeof(crypto::signature));
+
+    // 2. signature
+    std::string signature_string = decoded.substr(decoded.size() - sizeof(crypto::signature));
+    const crypto::signature &signature = *(const crypto::signature *)(signature_string.data());
+
+    // 3. kex round
+    std::size_t offset = sizeof(std::uint32_t);
+    std::string round_string = decoded.substr(0, offset);
+    const char *kex_round_bytes = round_string.c_str();
+    m_kex_round = 0;
+
+    for (std::size_t i{0}; i < 4; ++i)
+    {
+      std::uint32_t byte{static_cast<std::uint32_t>(kex_round_bytes[i])};
+      m_kex_round ^= (byte << i*8);
+    }
+    CHECK_AND_ASSERT_THROW_MES(m_kex_round > 0, "Invalid kex message round (must be > 0).");
+
+    // 4. signing public key
+    m_signing_pubkey = *(const crypto::public_key*)(decoded.data() + offset);
+    offset += sizeof(crypto::public_key);
+
+    // must be in prime subgroup
+    CHECK_AND_ASSERT_THROW_MES(m_signing_pubkey != crypto::null_pkey && m_signing_pubkey != rct::rct2pk(rct::identity()),
+      "Message signing pubkey was invalid.");
+    CHECK_AND_ASSERT_THROW_MES((rct::scalarmultKey(rct::pk2rct(m_signing_pubkey), rct::curveOrder()) == rct::identity()),
+      "Message signing pubkey was not in prime subgroup.");
+
+    // 5. msg pub keys & priv key
+    // note: sizeof(crypto::secret_key) == sizeof(crypto::public_key)
+    CHECK_AND_ASSERT_THROW_MES((decoded.size() - offset - sizeof(crypto::signature)) % sizeof(crypto::public_key) == 0,
+      "Multisig kex msg malformed.");
+    std::size_t num_keys = (decoded.size() - offset - sizeof(crypto::signature)) / sizeof(crypto::public_key);
+    m_msg_pubkeys.clear();
+    m_msg_pubkeys.reserve(num_keys);
+
+    if (m_kex_round == 1)
+    {
+      // in round 1 the message stores a private ancillary key component for the multisig account
+      // that will be shared by all participants (e.g. a shared private view key)
+      CHECK_AND_ASSERT_THROW_MES(num_keys == 1,
+        "Multisig kex msg malformed (round 1 message should only contain one private key).");
+
+      m_msg_privkey = *(const crypto::secret_key*)(decoded.data() + offset);
+
+      CHECK_AND_ASSERT_THROW_MES(sc_check((const unsigned char*)&m_msg_privkey) == 0 &&
+        m_msg_privkey != crypto::null_skey, "Multisig kex msg malformed (round 1 msg's privkey was invalid).");
+    }
+    else
+    {
+      for (std::size_t key_index{0}; key_index < num_keys; ++key_index)
```

Comment:
If the `src/serialization` change is too problematic (because of this divergent path here), I would do a `resize` instead of a `reserve`, with a single `memcpy` for the entire vector, then loop over for the checks. It makes the loops easier as `offset` no longer needs to be updated, etc.